appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopeContext.java

```diff
@@ -118,8 +118,35 @@ public boolean isActive() {
         }
     }
     
+    /**
+     * Get the most appropriate name for a bean. First checks the `@Clustered`
+     * annotation key name property, then the bean EL name, then the bean class
+     * name.
+     * 
+     * @param bean       the bean to reference.
+     * @param annotation the Clustered annotation to reference.
+     * @throws IllegalArgumentException if no name can be found for the bean.
+     */
     static <TT> String getBeanName(Bean<TT> bean, Clustered annotation) {
-        return annotation.keyName().isEmpty()? bean.getName() : annotation.keyName();
+        // Check the annotation keyName
+        String beanName = annotation.keyName();
+        if (beanName != null && !beanName.isEmpty()) {
+            return beanName;
+        }
+
+        // Check the CDI bean name
+        beanName = bean.getName();
+        if (beanName != null && !beanName.isEmpty()) {
+            return beanName;
+        }
+
+        // Check the bean class name
+        beanName = bean.getBeanClass().getName();
+        if (beanName != null && !beanName.isEmpty()) {
+            return beanName;
+        }
+
```

Comment:
Is there any way to achieve the same result without using the exact same if statement 3 times? I know it might not be but it would certainly be nicer in my opinion

Reply:
There is with either something like:

~~~
Stream.of(annotation.keyName(), bean.getName(), bean.getBeanClass().getName())
                .filter(obj -> obj != null && !obj.toString().isEmpty())
                .findFirst()
                .orElseThrow(new IllegalArgumentException("Could not find the name for bean: " + bean.toString()));
~~~

or:

~~~
private static <T> T firstNonNull(T ...items) {
        for(T i : items) if(i != null && !i.toString().isEmpty()) return i;
        throw new IllegalArgumentException("Could not find the name for bean: " + bean.toString());
}
~~~

IMO they both reduce clarity when debugging, although the second is somewhat more palatable. What do you think?

Reply:
I think I'd prefer the latter option to be used. I agree that it may make debugging slightly more involved, although the reduced code repetition will certainly aid in prevention of potential human error in further changes (e.g. someone accidentally changing only one of the if statements) - and I feel this option is still perfectly human readable

Reply:
I like the elegance of the stream, and feel like it could be made more readable, maybe by seperating the creation of the list before opening the stream. I like the second solution also, but I see where you're coming from when you mention clarity.